posts/inside-rust/2022-04-20-libs-aspirations.md

```diff
@@ -0,0 +1,209 @@
+  to live outside of the `rust-lang/rust` repository for less popular platforms.
+- A better way to allow only parts of `std` to be available, depending on the platform.
+  For example, a `where Platform: Unix` bound, or something like a [`#[cfg]` portibility lint](https://rust-lang.github.io/rfcs/1868-portability-lint.html).
+- Make the standard library more modular, allowing to disable e.g. floating point support
+  or file system support on certain platforms.
+
+### Improving and adding new APIs
+
+A main focus of the library team is and will always be the public interface of the standard library.
+As of last year, we even have a separate team to make the final calls for API changes and additions:
+the [library API team](https://www.rust-lang.org/governance/teams/library#Library%20API%20team).
+
+Rust purposely has a minimal standard library. Lots of commonly used functionality is
+found in other crates in the ecosystem, rather than the standard library.
+
+Where exactly we draw the line between things that should and shouldn't go in the standard library
+can be tricky to define and is somewhat flexible, but there are a few categories we're most interested in.
+
+**Ergonomics**
+
+A lot of additions to the standard library are very small ones that increase ergonomics.
+Quite often, these are things that were already possible in some way, just not in an ergonomic way.
+Some recent examples are:
+
+- `abs_diff()`
+- `Path::is_symlink`
+- `iter::from_fn`
+- `NonZero*::saturating_add`
+
+While we always have to consider the trade-off for niche features to the already large interface
+on some types and traits, additions like these continue to happen regularly.
+
+**Standardizing some bigger features the ecosystem needs**
+
+As Rust grows into new territories, there is more and more a need
+for certain features to be included in the standard library.
+This is especially true for things where a consistent, standard, interface is
+important.
+Some of the bigger examples are:
+
+- Async traits and functions
+- Allocators and fallible allocation
+- Error and panic handling
+- Portable SIMD
+
+**Reducing and improving unsafe code**
+
+By providing the right low level APIs and abstractions, we can greatly minimize
+the amount of complex unsafe code that users need to write. Tools like
+`MaybeUninit` guide users to correct unsafe code that's easy to follow and
+prove correct. Even better, some APIs can entirely remove the need for unsafe
+code in many situations.
+This includes situations where users tend to reach for `unsafe` for performance reasons.
+
+- `std::arch`
+- `std::simd`
+- Scoped threads
+- More atomic primitives
+- '`Iterator`' with static length for arrays
+- Improving `MaybeUninit` and related methods
+- Extending `NonNull` and pointer methods
+- A more complete interface to `OsString`, `Path`, and `CString`
+- Documentation for `Pin` and other 'unsafe' types
+- File descriptors and handles (`OwnedFd`, etc.)
```

Comment:
Safe transmute would be another addition here, as it's a common case for people reaching for unsafe. Sadly I think it's stalled.

I guess this list doesn't need to have everything in it.

Reply:
Safe transmute was mentioned in the [compiler team ambitions post](https://blog.rust-lang.org/inside-rust/2022/02/22/compiler-team-ambitions-2022.html#expressiveness-initiatives--). I believe jswrenn plans to be working on it extensively this summer.